public/app/plugins/datasource/prometheus/result_transformer.ts

```diff
@@ -1,234 +1,262 @@
-    const prometheusResult = response.data.data.result;
-
-    if (options.format === 'table') {
-      return [
-        this.transformMetricDataToTable(
-          prometheusResult,
-          options.responseListLength,
-          options.refId,
-          options.meta,
-          options.valueWithRefId
-        ),
-      ];
-    } else if (prometheusResult && options.format === 'heatmap') {
-      let seriesList: TimeSeries[] = [];
-      for (const metricData of prometheusResult) {
-        seriesList.push(this.transformMetricData(metricData, options, options.start, options.end));
-      }
-      seriesList.sort(sortSeriesByLabel);
-      seriesList = this.transformToHistogramOverTime(seriesList);
-      return seriesList;
-    } else if (prometheusResult) {
-      const seriesList: TimeSeries[] = [];
-      for (const metricData of prometheusResult) {
-        if (response.data.data.resultType === 'matrix') {
-          seriesList.push(this.transformMetricData(metricData, options, options.start, options.end));
-        } else if (response.data.data.resultType === 'vector') {
-          seriesList.push(this.transformInstantMetricData(metricData, options));
-        }
-      }
-      return seriesList;
-    }
+import {
+  ArrayVector,
+  DataFrame,
+  Field,
+  FieldType,
+  formatLabels,
+  TIME_SERIES_TIME_FIELD_NAME,
+  TIME_SERIES_VALUE_FIELD_NAME,
+} from '@grafana/data';
+import { FetchResponse } from '@grafana/runtime';
+import templateSrv from 'app/features/templating/template_srv';
+import {
+  PromDataSuccessResponse,
+  PromMatrixData,
+  PromMetric,
+  PromValue,
+  PromVectorData,
+  TransformOptions,
+} from './types';
+
+type MatrixOrVectorResult = PromMatrixData['result'][0] | PromVectorData['result'][0];
+
+export function transform(response: FetchResponse<PromDataSuccessResponse>, options: TransformOptions) {
+  const prometheusResult = response.data.data;
+
+  if (!prometheusResult.result) {
     return [];
   }
 
-  transformMetricData(metricData: any, options: any, start: number, end: number): TimeSeries {
-    const dps = [];
-    const { name, labels, title } = this.createLabelInfo(metricData.metric, options);
+  if (prometheusResult.resultType === 'scalar') {
+    return [
+      {
+        meta: options.meta,
+        refId: options.refId,
+        length: 1,
+        fields: [getTimeField([prometheusResult.result]), getValueField([prometheusResult.result])],
+      },
+    ];
+  }
 
-    const stepMs = parseFloat(options.step) * 1000;
-    let baseTimestamp = start * 1000;
+  if (options.format === 'table') {
+    const tableData = transformMetricDataToTable(prometheusResult.result, options);
+    return [tableData];
+  }
 
-    if (metricData.values === undefined) {
-      throw new Error('Prometheus heatmap error: data should be a time series');
-    }
+  const dataFrame: DataFrame[] = [];
+
+  prometheusResult.result.forEach((data: MatrixOrVectorResult) => dataFrame.push(transformToDataFrame(data, options)));
```

Comment:
I feel like this is not super clear mainly in the sense that which branch is exclusive with what and what we are actually checking. Something like this, just adding a few comments and maybe add indentation to mark a block:
```
if (resultType === 'scalar') { transform and return }
if (format === 'table') {
  return table
} else {
  // format is timeseries or heatmap I assume

  // In case of heatmap we process timeseries anyway and heatmap is then createad from that
  processTimeseries

  if (format === 'heatmap') {
    return heatmap
  }
}
```